src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -423,50 +424,97 @@ private TypeSymbol NarrowVariableAccessType(TypeValidatorConfig config, Variable
+        {
+            List<UnionTypeMemberViabilityInfo> candidacyEvaluations = new();
+
+            foreach (var candidateTargetType in targetType.Members)
+            {
+                if (!AreTypesAssignable(expressionType, candidateTargetType.Type))
+                {
+                    candidacyEvaluations.Add(new(candidateTargetType, null, ImmutableArray<IDiagnostic>.Empty));
+                    continue;
+                }
+
+                var candidateDiagnostics = ToListDiagnosticWriter.Create();
+                var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
+                var narrowed = candidateCollector.NarrowType(config, expression, expressionType, candidateTargetType.Type);
+                candidacyEvaluations.Add(new(candidateTargetType, narrowed, candidateDiagnostics.GetDiagnostics()));
             }
 
             var viableCandidates = candidacyEvaluations
-                .Select(c => c.Narrowed is {} Narrowed && !c.Diagnostics.OfType<ErrorDiagnostic>().Any() ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
+                .Select(c => c.Narrowed is {} Narrowed && !c.HasErrors ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
                 .WhereNotNull()
                 .ToImmutableArray();
 
             if (viableCandidates.Any())
             {
-                // It's unclear what we should do with warning and informational diagnostics if there's no path that's assignable without issue.
-                // Should we report only diagnostics raised by every viable candidate? Or report only those raised by a single candidate?
-                // Erring on the side of caution here and just reporting them all.
-                diagnosticWriter.WriteMultiple(viableCandidates.SelectMany(c => c.Diagnostics));
+                // It's unclear what we should do with warning and informational diagnostics. Should we report only the intersection of diagnostics raised by every viable candidate?
+                // Erring on the side of caution here and just reporting them all unless there are one or more candidates that are assignable w/o warning.
+                if (viableCandidates.All(c => c.Diagnostics.Any()))
+                {
+                    diagnosticWriter.WriteMultiple(viableCandidates.SelectMany(c => c.Diagnostics));
+                }
 
                 return TypeHelper.CreateTypeUnion(viableCandidates.Select(c => c.Type));
             }
 
             return ErrorType.Create(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(expressionType, targetType));
         }
 
-        private record UnionTypeMemberViabilityInfo(ITypeReference Type, TypeSymbol? Narrowed, IEnumerable<IDiagnostic> Diagnostics);
+        private record UnionTypeMemberViabilityInfo(ITypeReference Type, TypeSymbol? Narrowed, IEnumerable<IDiagnostic> Diagnostics)
```

Comment:
>UnionTypeMemberViabilityInfo [](http://example.com/codeflow?start=23&length=28)

Not really in scope of this PR, but would be good to add comments for the members of this record to document intent.